src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -715,90 +717,97 @@ event::ptr primitive_inst::update_weights() {
-    bool requires_reorder = weights_params.engine != kernel_selector::GenericKernelParams::Engine::NONE;
-
-    const auto weights_idx = _node->get_primitive()->input.size();
-    const auto original_weights_memory = dep_memory_ptr(weights_idx);
-    auto expected_layout = requires_reorder ? from_weights_tensor(weights_params.dest)
-                                            : original_weights_memory->get_layout();
-
-    // Set original patrial shape, because it may be lost during kernel_selector::weights_tensor -> layout conversion
-    expected_layout.set_partial_shape(original_weights_memory->get_layout().get_partial_shape());
-
-    if (requires_reorder && !_reordered_weights_cache.has(expected_layout)) {
-        GPU_DEBUG_PROFILED_STAGE_CACHE_HIT(false);
-        auto original_layout = original_weights_memory->get_layout();
-        auto& engine = _network.get_engine();
-
-        auto get_kernel_key = [&]() -> size_t {
-            auto seed = _node->get_primitive()->hash();
-            seed = hash_combine(seed, expected_layout.hash());
-            seed = hash_combine(seed, original_layout.hash());
-            return seed;
-        };
 
-        cldnn::kernel::ptr kernel = nullptr;
-        auto kernel_key = get_kernel_key();
-        auto& cache = get_network().get_in_mem_kernels_cache();
-        if (cache.has(kernel_key)) {
-            GPU_DEBUG_TRACE_DETAIL << id() << ": reorder weights (cached) from " << original_layout.to_short_string()
-                                   << " to " << expected_layout.to_short_string() << std::endl;
+    auto weights_idx = _node->get_primitive()->input.size();
+    auto original_weights_memory = dep_memory_ptr(weights_idx);
+    auto original_layout = original_weights_memory->get_layout();
+
+    if (weights_params.engine == kernel_selector::GenericKernelParams::Engine::NONE) {
+        // If kernel doesn't says that it doesn't require weights reorder, but weights were reordered previously, then
+        // incorrect memory buffer may be assigned, so reset cached weights for such case
+        _reordered_weights_cache.add(original_weights_memory->get_layout(), original_weights_memory);
+    } else {
+        auto expected_layout = from_weights_tensor(weights_params.dest);
+        // Set original patrial shape, because it may be lost during kernel_selector::weights_tensor -> layout conversion
+        expected_layout.set_partial_shape(original_weights_memory->get_layout().get_partial_shape());
+        _impl_params->weights_layout = optional_layout(expected_layout);
+
+        if (_reordered_weights_cache.has(expected_layout)) {
             GPU_DEBUG_PROFILED_STAGE_CACHE_HIT(true);
-            kernel = cache.get(kernel_key);
+            GPU_DEBUG_TRACE_DETAIL << id() << ": reuse weights for " << expected_layout.to_short_string() << std::endl;
+            return nullptr;
+        } else if (original_layout.compatible(expected_layout)) {
```

Comment:
Maybe it would be better to check for compatibility not only original layout, but all other cached layout as well

Reply:
Yeah, that can be a further improvement